File: client/csharp/example/cli/Menu.cs
using System;
using System.Collections.Generic;

namespace example
{
  public delegate void MenuCallback();

  class MenuItem {
    public string key;
    public string text;
    public MenuCallback callback;
    public bool exit;

    public MenuItem(string menuKey, string menuText, MenuCallback menuCallback, bool menuExit) {
      key = menuKey;
      text = menuText;
      callback = menuCallback;
      exit = menuExit;
    }

    public void Show() {
      Console.WriteLine("({0}) {1}", key, text);
    }
  }

  class InputItem {
    public string text;
    public string defaultVal;
  }

  class Menu
  {
    private MenuItem[] items;

    public Menu(MenuItem[] menuItems) {
      items = menuItems;
    }

    public void Show(string title) {
      while(true) {
        Console.WriteLine();
        Console.WriteLine("===== {0} =====", title);
        Console.WriteLine();

        for(int i = 0; i < items.Length; i++) {
          items[i].Show();
        }

        Console.WriteLine();
        Console.Write(">>>>> Select a menu: ");

        string input = Console.ReadLine().Trim();

        for(int i = 0; i < items.Length; i++) {
          if(input.Equals(items[i].key)) {
            if(items[i].callback != null) {
              items[i].callback();
            }

            if(items[i].exit) {
              return;
            } 
          }
        }
      }
    }

    public static string[] GetUserInput(InputItem[] items) {
      List<string> userInputs = new List<string>();

      for(int i = 0; i < items.Length; i++) {
        Console.Write(">> {0}: ", items[i].text);
        string input = Console.ReadLine().Trim();

        if(input.Equals("")) {
          input = items[i].defaultVal;
        }

        userInputs.Add(input);
      }

      return userInputs.ToArray();
    }

    public static uint[] GetDeviceIDs() {
      List<uint> deviceIDs = new List<uint>();

      while(true) {
        Console.Write(">> Enter the device ID (Press just ENTER if no more device): ");
        string input = Console.ReadLine().Trim();

        if(input.Equals("")) {
          break;
        }

        uint devID = 0;
        if(!UInt32.TryParse(input, out devID)) {
          Console.WriteLine("Invalid device ID: {0}", input);
          return null;
        }

        deviceIDs.Add(devID);
      }
      
      return deviceIDs.ToArray();
    }    
  }
}

File: client/node/example/cli/menu.js
const readlineSync = require('readline-sync');
const readline = require('readline');

class MenuItem {
  constructor(key, text, callback, exitMenu) {
    this.key = key;
    this.text = text;
    this.callback = callback;
    this.exitMenu = exitMenu;
  }
}

async function showMenu(title, menuItems) {
  var exitMenu = false;

  console.log(`\n===== ${title} =====\n`);

  for(let i = 0; i < menuItems.length; i++) {
    console.log(`(${menuItems[i].key}) ${menuItems[i].text}`);
  }

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  rl.question('\n>>>>> Select a menu: ', async (input) => {
    rl.close();

    var validMenu = false;
    var exitMenu = false;

    for(let i = 0; i < menuItems.length; i++) {
      if(input == menuItems[i].key) {
        validMenu = true;
        if(menuItems[i].callback != null) {
          if(menuItems[i].exitMenu) {
            menuItems[i].callback();
          } else {
            try {
              await menuItems[i].callback();
            } catch (err) {
              console.error(err);
            }
          } 
        }

        if(menuItems[i].exitMenu) {
          exitMenu = true;
        }

        break;
      }
    }

    if(!validMenu) {
      console.error(`Invalid menu: ${input}`);
    }

    if(!exitMenu) {
      showMenu(title, menuItems);
    }
  });
}

class UserInput {
  constructor(text, defaultVal) {
    this.text = text;
    this.defaultVal = defaultVal;
  }
}

function getUserInput(userInputs) {
  var inputVals = []

  for(let i = 0; i < userInputs.length; i++) {
    var input = readlineSync.question(`>> ${userInputs[i].text}: `);
    if(input == '' && userInputs[i].defaultVal != '') {
      inputVals.push(userInputs[i].defaultVal);
    } else {
      inputVals.push(input);
    }
  }

  return inputVals;
}


function getDeviceIDs() {
  var deviceIDs = [];

  while(true) {
    var input = readlineSync.question("Enter the device ID (Press just ENTER if no more device): ");

    if(input == '') {
      break;
    }

    var devID = Number(input);

    if(isNaN(devID)) {
      console.error(`Invalid device ID: ${input}`);
      break;
    }

    deviceIDs.push(devID);
  }

  return deviceIDs;
}

module.exports.MenuItem = MenuItem;
module.exports.showMenu = showMenu;
module.exports.UserInput = UserInput;
module.exports.getUserInput = getUserInput;
module.exports.getDeviceIDs = getDeviceIDs;